File: hashing/mapping-st/347.py
from collections import Counter

class Solution:
    def topKFrequent(self, nums, k):
        n = len(nums)
        mp = Counter(nums)
        bucket = [[] for _ in range(n + 1)]

        for element, freq in mp.items():
            bucket[freq].append(element)

        result = []
        for i in range(n, 0, -1):
            if not bucket[i]:
                continue
            while bucket[i] and k > 0:
                result.append(bucket[i].pop())
                k -= 1
        return result

File: hashing/mapping-st/347.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
    const n = nums.length
    const mp = new Map()
    for (const num of nums) {
        mp.set(num, (mp.get(num) || 0) + 1)
    }
    const bucket = Array.from({ length: n + 1 }, () => [])
    for (const [element, freq] of mp.entries()) {
        bucket[freq].push(element)
    }
    const result = []
    for (let i = n; i > 0; i--) {
        if (bucket[i].length === 0) continue
        while (bucket[i].length > 0 && k > 0) {
            result.push(bucket[i].pop())
            k--
        }
    }
    return result
};